File: CA4/Q4.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, u):
        if self.parent[u] != u:
            self.parent[u] = self.find(self.parent[u])
        return self.parent[u]

    def union(self, u, v):
        root_u = self.find(u)
        root_v = self.find(v)
        if root_u != root_v:
            if self.rank[root_u] > self.rank[root_v]:
                self.parent[root_v] = root_u
            elif self.rank[root_u] < self.rank[root_v]:
                self.parent[root_u] = root_v
            else:
                self.parent[root_v] = root_u
                self.rank[root_u] += 1
            return True
        return False

def kruskal_with_mandatory_edges(n, edges, mandatory_edges):
    uf = UnionFind(n)
    mst_cost = 0
    added_edges = 0

    # Add mandatory edges first
    for u, v, w in mandatory_edges:
        if uf.union(u, v):
            mst_cost += w
            added_edges += 1
        else:
            return float('inf')  # Cycle detected, invalid MST

    # Add remaining edges to complete MST
    for u, v, w in edges:
        if added_edges == n - 1:
            break
        if uf.union(u, v):
            mst_cost += w
            added_edges += 1

    return mst_cost if added_edges == n - 1 else float('inf')

def can_include_bridges(n, m, edges, queries):
    edges = sorted(edges, key=lambda x: x[2])  # Sort edges by weight
    results = []

    for query in queries:
        k_i = query[0]
        requested_bridges_indices = query[1:k_i + 1]

        # Prepare mandatory edges and remaining edges
        mandatory_edges = [edges[idx - 1] for idx in requested_bridges_indices]
        remaining_edges = [edges[i] for i in range(m) if (i + 1) not in requested_bridges_indices]

        # Calculate MST cost with mandatory edges
        new_mst_cost = kruskal_with_mandatory_edges(n, remaining_edges, mandatory_edges)

        # Check if MST is valid and matches the initial MST
        if new_mst_cost != float('inf'):
            results.append("YES")
        else:
            results.append("NO")

    return results

# Input Handling
def main():
    import sys
    input = sys.stdin.read
    data = input().split()
    index = 0

    # Read number of islands and bridges
    n, m = int(data[index]), int(data[index + 1])
    index += 2

    edges = []

    # Read the bridges information
    for _ in range(m):
        u, v, w = int(data[index]), int(data[index + 1]), int(data[index + 2])
        edges.append((u - 1, v - 1, w))  # Convert to zero-indexed
        index += 3

    # Read number of queries
    q = int(data[index])
    index += 1

    queries = []

    # Read each query
    for _ in range(q):
        query_data = list(map(int, data[index:index + 1 + int(data[index])]))
        queries.append(query_data)
        index += 1 + int(data[index])

    # Function Call to get results
    results = can_include_bridges(n, m, edges, queries)

    # Print results for each query
    sys.stdout.write("\n".join(results) + "\n")

if __name__ == "__main__":
    main()

File: CA4/Q4.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <set>
#include <algorithm>
#include <limits> // For INT_MAX

using namespace std;

class UnionFind {
public:
    vector<int> parent, rank;

    UnionFind(int n) {
        parent.resize(n);
        rank.resize(n, 0);
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
        }
    }

    int find(int u) {
        if (parent[u] != u) {
            parent[u] = find(parent[u]); // Path compression
        }
        return parent[u];
    }

    bool unionSets(int u, int v) {
        int root_u = find(u);
        int root_v = find(v);
        if (root_u != root_v) {
            if (rank[root_u] > rank[root_v]) {
                parent[root_v] = root_u;
            } else if (rank[root_u] < rank[root_v]) {
                parent[root_u] = root_v;
            } else {
                parent[root_v] = root_u;
                rank[root_u]++;
            }
            return true;
        }
        return false;
    }
};

int kruskalWithMandatoryEdges(int n, const vector<tuple<int, int, int>>& edges, const vector<tuple<int, int, int>>& mandatoryEdges) {
    UnionFind uf(n);
    int mstCost = 0;
    int addedEdges = 0;

    // Add mandatory edges first
    for (auto [u, v, w] : mandatoryEdges) {
        if (uf.unionSets(u, v)) {
            mstCost += w;
            addedEdges++;
        } else {
            return INT_MAX; // Return infinity if a cycle is detected
        }
    }

    // Add remaining edges to complete the MST
    for (auto [u, v, w] : edges) {
        if (addedEdges == n - 1) break; // Stop if we already have a spanning tree
        if (uf.unionSets(u, v)) {
            mstCost += w;
            addedEdges++;
        }
    }

    return (addedEdges == n - 1) ? mstCost : INT_MAX;
}

vector<string> canIncludeBridges(int n, int m, vector<tuple<int, int, int>>& edges, vector<vector<int>>& queries) {
    vector<string> results;
    sort(edges.begin(), edges.end(), [](auto& a, auto& b) {
        return get<2>(a) < get<2>(b); // Sort edges by weight
    });

    // Compute the initial MST cost with all edges
    int initialMstCost = kruskalWithMandatoryEdges(n, edges, {});

    for (auto& query : queries) {
        int k_i = query[0];
        vector<tuple<int, int, int>> mandatoryEdges;
        set<int> requestedIndices(query.begin() + 1, query.end());

        // Create a list of mandatory edges
        for (int idx : requestedIndices) {
            mandatoryEdges.push_back(edges[idx - 1]); // Convert 1-indexed to 0-indexed
        }

        // Create a list of remaining edges
        vector<tuple<int, int, int>> remainingEdges;
        for (int i = 0; i < m; ++i) {
            if (requestedIndices.find(i + 1) == requestedIndices.end()) {
                remainingEdges.push_back(edges[i]);
            }
        }

        // Run Kruskal's algorithm
        int newMstCost = kruskalWithMandatoryEdges(n, remainingEdges, mandatoryEdges);

        // Compare costs with the initial MST cost
        results.push_back((newMstCost == initialMstCost) ? "YES" : "NO");
    }

    return results;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<tuple<int, int, int>> edges;
    for (int i = 0; i < m; ++i) {
        int u, v, w;
        cin >> u >> v >> w;
        edges.emplace_back(u - 1, v - 1, w); // Convert to zero-indexed
    }

    int q;
    cin >> q;

    vector<vector<int>> queries(q);
    for (int i = 0; i < q; ++i) {
        int k;
        cin >> k;
        queries[i].resize(k + 1);
        queries[i][0] = k;
        for (int j = 1; j <= k; ++j) {
            cin >> queries[i][j];
        }
    }

    vector<string> results = canIncludeBridges(n, m, edges, queries);
    for (const string& res : results) {
        cout << res << "\n";
    }

    return 0;
}
